--- lab2/kern/debug/kdebug.c	2021-11-07 14:43:58.974120722 +0800
+++ lab1/kern/debug/kdebug.c	2021-11-07 14:11:46.526075562 +0800
@@ -3,10 +3,7 @@
 #include <stab.h>
 #include <stdio.h>
 #include <string.h>
-#include <sync.h>
 #include <kdebug.h>
-#include <kmonitor.h>
-#include <assert.h>
 
 #define STACKFRAME_DEPTH 20
 
@@ -305,5 +302,19 @@
       *           NOTICE: the calling funciton's return addr eip  = ss:[ebp+4]
       *                   the calling funciton's ebp = ss:[ebp]
       */
+    uint32_t ebp = read_ebp(), eip = read_eip();
+    int i,j;
+    for (i = 0; ebp != 0 && i < STACKFRAME_DEPTH; i++) {
+        cprintf("ebp:%08x, eip:%08x ", ebp, eip);
+        for (j = 0; j < 4; j++) {
+            cprintf("args[%d]:%08x ", j, *((uint32_t*)ebp + 2 + j));
+        }
+        cprintf("\n");
+        print_debuginfo(eip-1);
+        eip = *((uint32_t*)ebp + 1);
+        ebp = *((uint32_t*)ebp);
+        //eip = ((uint32_t*)ebp)[1];
+        //ebp = ((uint32_t*)ebp)[0];
+    }
 }
 
